### ISSUE TITLE
Problems: Redundancy between docs.webkit.org and wiki; user confusion; PR costs. Solution: Centralize all doc markdown sources in a single git repo, serve both docs.webkit.org and wiki from same sources

### ISSUE THREAD

**Issue by sideshowbarker** [WebKit/Documentation#88]
## Problems

- Some docs content in the wiki is similar/redundant with docs.w3.org content.
- Having similar docs in multiple places is confusing for users. Which docs source is canonical — which should they use/trust?
- Some (potential) contributors reluctant/unwilling to do PRs to add/change docs.webkit.org content, due to time/trouble costs.
- Many (most) contributors are more likely to contribute to the wiki, because wiki contributing is easier + quicker (lower costs).
- Without totally disabling the wiki, there’s no way to _prevent_ new wiki content redundant with existing docs.webkit.org content.

## Solution

- Take all existing content from both the wiki and from docs.webkit.org, move/merge/centralize it into a single git/GitHub repo.
- Serve/deploy/publish centralized docs content to both docs.webkit.org and to the wiki as the single “canonical” project docs.

## Rationale

A single centralized/canonical repo for the docs:

- Eliminates all current redundancy.
- Avoids causing user confusion about which content is most up-to-date and canonical — because it’d all be the same content.
- Makes contributing as easy as possible for all potential contributors; lowest barriers, lowest time/trouble costs.
- Wiki contributors get the new benefit of their contributions automatically showing up at docs.webkit.org.
- Gives potential contributors significantly more incentive to take the time to contribute quality docs content.
- Gives happiness even to users who hate wikis: they can instead commit/push directly to https://github.com/WebKit/WebKit.wiki.git if they prefer.

### Questions/Arguments against + answers/rebuttals for those

Q: If all content sources are moved to the wiki, what use would anybody then have for docs.webkit.org?
A: Good question. One good answer is that docs.webkit.org does already exist and is a cool URL and [Cool URIs don't change](https://www.w3.org/Provider/Style/URI). But also, docs.webkit.org gives the project a way to brand/theme/skin the presentation of the docs in whatever way the project chooses — rather than being restricted to the very simple/limited common theme/branding that GitHub’s wiki engine produces.

Q: Somebody would need to do the actual work of setting it all up.
A: I’d be personally willing to put time and work into making it happen.

## Implementation notes

Here are some details about how the centralized solution should be implemented/deployed:

### In practice currently, the repo for the GitHub wiki would necessarily need to be the canonical repo

The wiki is backed by a GitHub repo that can be cloned from https://github.com/WebKit/WebKit.wiki.git. The wiki provides a frontend for editing the sources in that git repo, and also serves as the markdown->HTML publishing engine and publishing server for the content.

There is currently no way to associate a different/arbitrary git backing repo with GitHub wiki — so that necessarily means the existing https://github.com/WebKit/WebKit.wiki.git repo for the GitHub wiki would be the repo that becomes the canonical repo.

### All existing docs.webkit.org docs sources could be moved/merged into the git repo for the wiki

Because in practice the https://github.com/WebKit/WebKit.wiki.git repo necessarily needs to be the on that becomes the canonical repo, then that also necessarily means the existing docs.webkit.org sources would need to be moved/migrated (back) to the wiki.

### Make the entire docs/ subdirectory into a git submodule

The most-straightforward way to implement the proposed solution would be to change the entire existing https://github.com/WebKit/Documentation/tree/main/docs directory into just being a git submodule from the https://github.com/WebKit/WebKit.wiki.git wiki repo.

### For better docs.webkit.org site navigation, etc., pull in more mkdocs modules or move to new tool

Pulling content from the wiki into docs.webkit.org introduces some additional complications around page-titling and sorting and such that would benefit from having some finer control over navigation features for the site.

While it’s true that the core mkdocs engine itself does not provide very good control over site navigation and page sorting in navigation, etc. — there are some third-party plug-ins that do much more. Probably the best one [mkdocs-literate-nav](https://oprypin.github.io/mkdocs-literate-nav/)

But it could also be worth investigating whether it would be productive to move to another tool that’s more closely suited to publish-a-website-from-a-GitHub-wiki needs.

**Comment by cdumez** [comment]
I do not want for docs to be in a repository. Keeping docs up to date is already hard enough, if we add extra friction for updating (forking & making a commit / PR) then it will only be worse. For this reason, I want a wiki and I am using the wiki.

**Comment by cdumez** [comment]
> I do not wish for docs to be in a repository. Keeping docs up to date is already hard enough, if we add extra friction for updating (forking & making a commit / PR) then it will only be worse. For this reason, I want a wiki and I am using the wiki.

As a result, this proposed solution does not sound good to me:
> Take all existing content from both the wiki and from docs.webkit.org, move/merge/centralize it into a single git/GitHub repo.

**Comment by sideshowbarker** [comment]
@cdumez Understood — but to be clear: The wiki docs are _already_ in a repository, at https://github.com/WebKit/WebKit.wiki.git.

It’s just that when you make changes to the wiki via the wiki editing frontend, what happens on the backend is that you cause commits to be made to that git repository.

> As a result, I would object to this proposed solution:
> 
> > Take all existing content from both the wiki and from docs.webkit.org, move/merge/centralize it into a single git/GitHub repo.

To be more clear about my proposal and to put it in terms of the wiki: What I’m proposing is:
- The wiki content stays exactly where it is, and you continue to be able to edit it exactly as you do now.
**So, no changes at all in docs workflow for you.**
- The current contents of the repo for _docs.webkit.org_ are what get moved/merged into the wiki
So you’d then additionally be able to change/fix to any of _that_ content directly at the wiki, **instead of doing PRs**.
- To make docs changes, **you’d no longer need to do _any_ PRs _anywhere_**  — **you just do it all through the wiki.**

And please see that I’ve now added the above list as as **Goals** section in the issue description, and I’ve updated the title to:

> _Solution: Move/centralize all doc sources **into the wiki**, serve both docs.webkit.org and wiki from same sources_

Does that all make the proposal now sound more acceptable?

**Comment by mdubet** [comment]
> > @cdumez Understood — but to be clear: The wiki docs are _already_ in a repository, at [WebKit/WebKit.wiki.git](https://github.com/WebKit/WebKit.wiki.git).
> > It’s just that when you make changes to the wiki via the wiki editing frontend, what happens on the backend is that you cause commits to be made to that git repository.
> > > As a result, I would object to this proposed solution:
> > > > Take all existing content from both the wiki and from docs.webkit.org, move/merge/centralize it into a single git/GitHub repo.
> > 
> > 
> > To be more clear about my proposal and to put it in terms of the wiki: What I’m proposing is:
> > 
> > * The wiki content stays exactly where it is, and you continue to be able to edit it exactly as you do now.
> >   **So, no changes at all in docs workflow for you.**
> > * The current contents of the repo for _docs.webkit.org_ are what get moved/merged into the wiki
> >   So you’d then additionally be able to change/fix to any of _that_ content directly at the wiki, **instead of doing PRs**.
> > * To make docs changes, **you’d no longer need to do _any_ PRs _anywhere_**  — **you just do it all through the wiki.**
> > 
> > And please see that I’ve now added the above list as as **Goals** section in the issue description, and I’ve updated the title to:
> > > _Solution: Move/centralize all doc sources **into the wiki**, serve both docs.webkit.org and wiki from same sources_
> > 
> > 
> > Does that all make the proposal now sound more acceptable?
> 
> To me yes. I agree it'd be great to have documentation in a central place, as long as I keep wiki editing.

FWIW, it's possible to edit the documentation repository directly from the web browser in markdown (but you don't get to see the html result) example : https://github.com/WebKit/Documentation/edit/main/docs/index.md

**Comment by sideshowbarker** [comment]
> > I do not wish for docs to be in a repository. Keeping docs up to date is already hard enough, if we add extra friction for updating (forking & making a commit / PR) then it will only be worse. For this reason, I want a wiki and I am using the wiki.
> 
> FWIW, it's possible to edit the documentation repository directly from the web browser in markdown… example : [WebKit/Documentation/edit/main/docs/index.md](https://github.com/WebKit/Documentation/edit/main/docs/index.md)

@mdubet Yes, thanks much for pointing this out. Also worth mentioning here: all the https://docs.webkit.org docs have an _“Edit this page”_ pencil icon which, if you click it, takes you to the GitHub UI for editing the doc directly in your browser. 

But that said, the repo settings as currently configured have the main branch set to be a protected branch — so, once you make edits to the page, you can’t land (commit) them right away, but instead need to click a button to raise a PR.

However, it’s completely possible to change those settings so that project members with commit/push privileges actually _could_ commit directly to the main branch without needing to raise PRs. Or else, the settings could be updated such that the branch which changes from the GitHub UI would get landed (committed in) ine would be a _”staging”_ (or something) branch that a project owner subsequently would merge into the main branch (for publishing to docs.webkit.org).

> (but you don't get to see the html result)

True — but there _is_ also a **`Preview`** button which when you click on it shows you a preview of _a_ rendered version of the markdown — and that rendered preview is almost exactly like what you when editing a page in the GitHub wiki.

So @cdumez, can you please take 5 minutes or so to try this:

1. Go to a page like https://github.com/WebKit/Documentation/edit/main/docs/Deep%20Dive/MemoryManagement.md and trying making some throwaway/test edits to it directly in your browser.
2. Push the  **`Preview`** button there to see rendered output of your changes right away.

…then, that’s basically just like editing a doc at the wiki, right?

The only difference is that to actually land your edits, you then need to push the green **`Commit Changes…`** button, 
when then _currently_ will cause a PR to be raised (and then you need to go through the PR review process to land it).

_However_, what’s possible is that the setting could be changed such that when you pushed that **`Commit Changes…`** button, your edits would land immediately — without you needing to go through the PR process.

_If_ the settings were changed such that you do that — that is, make docs edits directly in your browser and land them right way, just like you can now at the wiki — would that workflow work OK for you?

It would basically be just like how you make changes to the current wiki — you’d just be doing it in a different place.

**Comment by mcatanzaro** [comment]
> I do not wish for docs to be in a repository. Keeping docs up to date is already hard enough, if we add extra friction for updating (forking & making a commit / PR) then it will only be worse. For this reason, I want a wiki and I am using the wiki.

This used to be my opinion too -- I think I mentioned pretty much exactly that when docs.webkit.org was first proposed -- but now that the docs website exists it's kinda too late. A lot of effort went into this website, it does the job, and having one place for our docs is by far the most important consideration. E.g. we can't rightly expect new contributors to follow the [new smart pointer guidelines](https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines) because [our smart pointer documentation makes no mention of them](https://docs.webkit.org/Deep%20Dive/MemoryManagement.html#how-to-use-refptr-and-ref).

I would just add Chris directly to the project so he doesn't have to use pull requests. I think most developers are probably happy to use pull requests, but let's not require them.

The GitHub web UI workflow for committing to docs.webkit.org is _really_ close to GitHub wiki; you edit in Markdown, and have live preview. The only flaw I see is that docs pages currently link to the [toplevel directory](https://github.com/WebKit/Documentation/tree/main) instead of the actual page of the content on GitHub. That creates a small amount of friction that doesn't exist in a wiki, which would be nice to get rid of.

**Comment by mcatanzaro** [comment]
Also, since both projects use Markdown, moving content from wiki to docs website is mostly going to be a matter of Ctrl+C Ctrl+V. So that, at least, is not a very huge change.
